### PR TITLE
HPCC-12934 Optimize IF(a, b+c, b+d) to b + IF(a,c,d)

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jamienoss please review.
This has a very significant affect on a small subset of our archived queries, and a small effect on quite a large number of them

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| appendoptimize | Pass | Fail | Pass |
 
HPCC Stop: OK

**Comment by ghalliday** [comment]
Looks like I need to use & not +

**Comment by richardkchapman** [comment]
Sounds familiar!

**Comment by richardkchapman** [comment]
Is the optimization equally valid for the + case as it is for the & case?

**Comment by ghalliday** [comment]
Yes - I'm not changing the order.  If the expression if

    IF(a, b + c, b & d)

then it will not apply the transformation.

**Comment by jamienoss** [comment]
Given the semantics of '+'  within this context I think it is correct to apply this optimization for the above example as

``` IF(a, b + c, b & d) \equiv IF(a, b & c, b & d) ```

File: ecl/hql/hqlopt.cpp [hunk 1]
```diff
@@ -2195,19 +2259,27 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
...
     switch (op)
     {
     case no_if:
         {
             OwnedHqlExpr ret = optimizeIf(transformed);
+
+            //This won't split shared nodes, but one of the children may be shared - so processed here
+            if (!ret && transformed->isDataset())
+            {
+                //Convert IF(a, f(ds), g(ds)) to ds(IF(a,f,g)) - revisit since may increase dependencies
```
Comment (jamienoss) [inline ecl/hql/hqlopt.cpp]: Have you opened a Jira to revisit this?

**Comment by richardkchapman** [comment]
@ghalliday Waiting for your reply...

File: ecl/hql/hqlopt.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlopt.cpp]: There is already a jira open. (https://track.hpccsystems.com/browse/HPCC-13300)

**Comment by ghalliday** [comment]
It would be possible to do the optimization Jamie suggests, but as I discussed with him, I really don't think it is worth the complication.

**Comment by ghalliday** [comment]
@richardkchapman see replies

**Comment by richardkchapman** [comment]
Did you fix the test case?

@jamienoss have you finished reviewing this?

**Comment by ghalliday** [comment]
The test case was correct - so investigating if it is highlighting an issue.

**Comment by ghalliday** [comment]
It was showing that the flag to indicate ordering wasn't being preserved...  smoke test caught a bug.
Fix released.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| appendoptimize | Pass | Pass | Pass |
 
HPCC Stop: OK

File: ecl/hql/hqlopt.cpp (hunk 1 above)
Comment (jamienoss) [inline ecl/hql/hqlopt.cpp]: This optimization is given precedence, yet it is not mentioned in either the jira issue or the PR summary.

Comment (ghalliday) [inline ecl/hql/hqlopt.cpp]: Can you clarify your comment?

Comment (jamienoss) [inline ecl/hql/hqlopt.cpp]: Sorry, I see that the first optimization is not new code, where  I thought it was.

**Comment by jamienoss** [comment]
@richardkchapman  This is good to merge, however, I am concerned that IF(a, b+c, b+d) to b + IF(a,c,d) maybe misunderstood. I would add somewhere that '+' is not commutative where it maybe generally presumed to be.
i.e. IF(a, b+c, d+b) is not optimized.

**Comment by richardkchapman** [comment]
+ Meaning unordered concat should be commutative I would have thought - it's the ordered case ( & ) that is not?

**Comment by jamienoss** [comment]
I don't quite follow your comment.
'+' is commutative and '&' is not by definition (along with the added implementation detail that order within the sets is not disrupted).

So... IF(a,b+c,b+d) \equiveto IF(a,b+c,d+b) and therefore this should be optimized also.
I only mean that it should be documented somewhere, if not here that this is in fact the implementation of a subset of IF(a, b+c, b+d) to b + IF(a,c,d)  or IF(a, b+c, b+d) to b + IF(a,c,d) where '+' is considered non-commutative.

**Comment by jamienoss** [comment]
@ghalliday, @richardkchapman. In fact, this is the complete implementation of  IF(a, b&c, b+d) to b & IF(a,c,d).
Having re-read the above comments and the reg-test in this change why are we even mentioning '+'?

**Comment by richardkchapman** [comment]
We don't (and won't) document what is/is not optimized so that's a non-issue.

**Comment by richardkchapman** [comment]
This optimizes IF (a, b OP c, b OP d) for OP in [ +, &, && ]

**Pull request merged**